rust/connlib/shared/src/messages/gateway.rs

```diff
@@ -0,0 +1,141 @@
+//! Gateway related messages that are needed within connlib
+
+use ip_network::IpNetwork;
+use serde::Deserialize;
+
+use super::ResourceId;
+
+pub type Filters = Vec<Filter>;
+
+/// Description of a resource that maps to a DNS record.
+#[derive(Debug, Deserialize, Clone, PartialEq, Eq)]
+pub struct ResourceDescriptionDns {
+    /// Resource's id.
+    pub id: ResourceId,
+    /// Internal resource's domain name.
+    pub address: String,
+    /// Name of the resource.
+    ///
+    /// Used only for display.
+    pub name: String,
+
+    pub filters: Filters,
+}
+
+/// Description of a resource that maps to a CIDR.
+#[derive(Debug, Deserialize, Clone, PartialEq, Eq)]
+pub struct ResourceDescriptionCidr {
+    /// Resource's id.
+    pub id: ResourceId,
+    /// CIDR that this resource points to.
+    pub address: IpNetwork,
+    /// Name of the resource.
+    ///
+    /// Used only for display.
+    pub name: String,
+
+    pub filters: Filters,
+}
+
+/// Description of a resource that maps to a DNS record which had its domain already resolved.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ResolvedResourceDescriptionDns {
+    pub id: ResourceId,
+    /// Internal resource's domain name.
+    pub domain: String,
+    /// Name of the resource.
+    ///
+    /// Used only for display.
+    pub name: String,
+
+    pub addresses: Vec<IpNetwork>,
+
+    pub filters: Filters,
+}
+
+#[derive(Debug, Deserialize, Clone, PartialEq, Eq)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ResourceDescription<TDNS = ResourceDescriptionDns> {
+    Dns(TDNS),
+    Cidr(ResourceDescriptionCidr),
+}
+
+#[derive(Debug, Deserialize, Clone, Copy, PartialEq, Eq, Hash)]
+#[serde(tag = "protocol", rename_all = "snake_case")]
+pub enum Filter {
+    Udp(FilterInner),
+    Tcp(FilterInner),
+    Icmp,
+}
+
+#[derive(Debug, Deserialize, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct FilterInner {
```

Comment:
Do we strictly need this? Why not just make a struct-variant on `Filter` with the corresponding field names?